File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/FileInfo.kt
package com.jetbrains.signatureverifier.serialization.fileInfos

import kotlinx.serialization.Serializable
import java.nio.channels.SeekableByteChannel

@Serializable
sealed class FileInfo {
  abstract val fileMetaInfo: FileMetaInfo
  abstract val signedDataInfo: SignedDataInfo
  open fun modifyFile(stream: SeekableByteChannel) = fileMetaInfo.modifyFile(stream, signedDataInfo.toSignature())
}

File: net/JetBrains.Serialization/src/FileInfos/FileInfo.cs
using Newtonsoft.Json;

namespace JetBrains.Serialization.FileInfos;

[JsonObject(MemberSerialization.OptIn)]
public abstract class FileInfo
{
  [JsonProperty("fileMetaInfo")]
  public abstract IFileMetaInfo FileMetaInfo { get; }

  [JsonProperty("signedDataInfo")]
  public abstract SignedDataInfo SignedDataInfo { get; }

  public virtual void ModifyFile(Stream stream)
  {
    FileMetaInfo.ModifyFile(stream, SignedDataInfo.ToSignature());
  }
}